### PR TITLE
gitlab: use baseURL to match the correct OAuth context

### PR DESCRIPTION
**Pull request by unknwon**
Previously in https://github.com/sourcegraph/sourcegraph/pull/39840, we would always pick the first GitLab auth provider to obtain the OAuth context, which isn't a problem when there is only one GitLab auth provider configured (in most of the cases). However, this would give us incorrect OAuth context when there are more than one GitLab auth providers configured.

This PR fixes that by using the root URL for determining the correct auth provider to get OAuth context.

## Test plan

Unit tests

---

Followup of https://github.com/sourcegraph/sourcegraph/pull/39840#discussion_r951759885

### CODE REVIEW

File: internal/extsvc/gitlab/client.go [hunk 1]
```diff
@@ -435,16 +435,23 @@ func (e ProjectNotFoundError) NotFound() bool { return true }
...
+	baseURL.Path = "" // Cut out path like "/api/v4"
+	rootURL := strings.TrimSuffix(baseURL.String(), "/")
 	for _, authProvider := range conf.SiteConfig().AuthProviders {
 		if authProvider.Gitlab != nil {
 			p := authProvider.Gitlab
-
 			glURL := strings.TrimSuffix(p.Url, "/")
+			if glURL != rootURL {
+				continue
+			}
```
Comment (eseliger) [inline internal/extsvc/gitlab/client.go:453]: Is the baseURL always ensured to have a trimmed prefix as well? Slightly scared that there can be mismatches here that remain unnoticed and we just silently return `nil` 😬

**Review by eseliger — APPROVED** [review]
Thanks for the quick follow-up!

File: internal/extsvc/gitlab/client.go (hunk 1 above)
Comment (unknwon) [reply internal/extsvc/gitlab/client.go:453]: Hmm, "a trimmed prefix"?

Comment (unknwon) [reply internal/extsvc/gitlab/client.go:453]: Oh, the other way to do is something like `if strings.HasPrefix(baseURL.String(), glURL)`, is that what you're thinking of?

Comment (eseliger) [reply internal/extsvc/gitlab/client.go:453]: maybe? Or just compare URL.Host and URL.Port? not sure 🤔

Comment (unknwon) [reply internal/extsvc/gitlab/client.go]: OK, I made another mistake which makes in-place mutation `baseURL.Path = ""` which breaks tons of things 😂 Changed to use the prefix matching.

**Comment by pjlast** [comment]
I'm wondering what happens if there are two GitLab auth providers with the same baseURL? Are there checks that prevent this from happening?

**Comment by unknwon** [comment]
> I'm wondering what happens if there are two GitLab auth providers with the same baseURL? Are there checks that prevent this from happening?

That's an invalid configuration I think? At least I don't think we support that, because they'll have identical `serviceID` and would confuse other application logic we have (e.g. repo perms syncing). 🤔

**Comment by pjlast** [comment]
But nothing prevents it from actually being set up that way. Currently, for example, GitHub App auth provider is set up alongside the GitHub OAuth auth provider on s2. If this is truly undefined behaviour, then I think we should perhaps create a separate issue to implement checks that does not allow it. But will approve this one as is, as I think it is unrelated

**Review by pjlast — APPROVED** [review]

**Review by miveronese — APPROVED** [review]
good catch!!

**Pull request merged**